### ISSUE TITLE
Disposing a render target on a different thread in DesktopGL can cause an exception

### ISSUE THREAD

**Issue by Ellpeck** [MonoGame/MonoGame#7991]
This is a very rare issue that happens when calling `Dispose` on a `RenderTarget2D` or `RenderTargetCube` on a thread that isn't the UI thread, while also calling `Dispose` on the UI thread, likely introduced here: https://github.com/MonoGame/MonoGame/commit/a65158d4fb4646ce0f0484dd2153b50e4e7cdfa7#diff-46de2aa5f6c51e1f3254d0e88ea3776ae4417c8cec350afa0195730e0776d390R13.

It *appears* that this issue occurs because the `Dispose` method for the render target classes sets the `GraphicsDevice` to null, and when the `DisposeAction` is executed at a later point (due to waiting until it is on the UI thread), the `GraphicsDevice` has already been set to null and the following `NullReferenceException` is thrown:

```
System.NullReferenceException: Object reference not set to an instance of an object.
   at Microsoft.Xna.Framework.Graphics.RenderTarget2D.<>c.<.cctor>b__47_0(RenderTarget2D t)
   at Microsoft.Xna.Framework.Threading.StateActionHelper`1.Dequeue()
   at Microsoft.Xna.Framework.Threading.Run()
   at Microsoft.Xna.Framework.SdlGamePlatform.RunLoop()
   at Microsoft.Xna.Framework.Game.Run(GameRunBehavior runBehavior)
   at Program.<Main>$(String[] args) in C:\Users\me\Desktop\Test\Program.cs:line 3

```

This issue happens *exceptionally rarely* in a real-world scenario, where it's extremely difficult to reproduce, but I've created this extremely awful test game to somewhat consistently reproduce this issue:

```cs
public class Game1 : Game {

    private GraphicsDeviceManager graphics;
    private readonly RenderTarget2D[] targets = new RenderTarget2D[1000];

    public Game1() {
        this.graphics = new GraphicsDeviceManager(this);
        this.Content.RootDirectory = "Content";
        this.IsMouseVisible = true;
    }

    protected override void Initialize() {
        base.Initialize();
        for (var i = 0; i < 1000; i++)
            this.TestTarget(i);
    }

    protected override void Update(GameTime gameTime) {
        // dispose all targets when pressing (well, holding) D, which causes the Threading action to fail
        if (Keyboard.GetState().IsKeyDown(Keys.D)) {
            foreach (var target in this.targets)
                target.Dispose();
        }
        base.Update(gameTime);
    }

    private async void TestTarget(int index) {
        this.targets[index] = new RenderTarget2D(this.GraphicsDevice, 10, 10);
        await Task.Delay(1000 + 10 * index);
        Console.WriteLine($"Disposing target {index} on thread {Environment.CurrentManagedThreadId}");
        this.targets[index].Dispose();
    }

}
```

This test code just starts a bunch of asynchronous tasks, disposing render targets at a later date. However, it also allows pressing `D` to dispose all targets on the main thread, which (almost always) causes this issue to occur.

Lastly, I want to make clear that I don't know how *serious* this issue is, as disposing the render target from a different thread is arguably already a questionable thing to do, but disposing it a second time on the main thread is even more questionable. That being said, the `IDisposable` documentation explicitly mentions that calling `Dispose` multiple times on an object that implements it should be safe and not throw any exceptions, which is in line with all the other MonoGame disposables. 

#### What version of MonoGame does the bug occur on:
- MonoGame 3.8.1.303

#### What operating system are you using:
- Windows

#### What MonoGame platform are you using:

- DesktopGL

**Comment by Ellpeck** [comment]
I've recently learned that this behavior can not only cause an exception, but also cause a deadlock in the game when exiting sometimes. When creating a render target and then not disposing it manually, its finalizer will be called by the garbage collector. If this happens while or shortly before the game is exiting, the dispose action will be enqueued using `Threading.BlockOnUIThread`, which waits for a `ManualResetEventSlim` *indefinitely*, presumably due to the fact that the dispose queue is never accessed again when the game exits.

This is a screenshot taken from a running game that has been closed through the X button minutes before, and is now waiting indefinitely for this event to be invoked, causing it to be stuck in a running state.

![image](https://user-images.githubusercontent.com/5741138/230347571-e3a66f9b-4014-4011-9268-5b8d28f20a24.png)

**Comment by Ravendarke8** [comment]
Mg's desktopGL shouldn't, in no way, be considered multithreadable. Multithread rendering and deferred context is even outside of scope of DirectX mg (which has like almost of decade of difference modernity wise). So simply put: Don't.

**Comment by Ellpeck** [comment]
As I said at the top: I don't know how serious this issue is, but the fact that the `Dispose` implementation *itself* is clearly made to be some type of thread-safe, and so the fact that that implementation doesn't work correctly/consistently is something that I think should be flagged up and possibly fixed or addressed some other way.

**Comment by mrhelmut** [comment]
While OpenGL is absolutely not a multithread renderer (like any other MG backend) and must not be used in threads, it is true that MG has a mechanism to allow Texture and RenderTarget creation on threads, but not quite.

As you noticed, loading textures or RTs on a thread with DesktopGL, actually blocks the thread until the Draw() method ends and then executes the actual loading on the main thread before unblocking the thread. This was made to mimic Direct3D 9~11 behavior (which allows to load resources from threads), with the caveat of not actually doing the loading on a thread, which may or may not visibly impair the main thread.

This said, this totally sounds like a race condition in your thread with the same RT entering twice its dispose action.

**Comment by damian-666** [comment]
Thx for the detailed analysis.   Could this cause a process leak people see
in the dev studio debugger that they blame on Mg, on unhandled exception,
end debug session, is very annoying. I use a render target on my draw
thread to create sprites.  I don't explicitly dispose them bc there isn't
an idisposabe on the proxy api.  But I run a bk thread, get a leaked
process often.  Have to close developer studio,  find the process when it
finally appears in tak manager, kill it..     might it affect dx, and is
there a workaround  ?  Would be Just a one line fix seems like..  this
causal set of changes you pointed out was intended to help with reducing
garbage creation?  Seems to have introduced some regression..from two
things you mentioned.   One regression I got around using the nullable
warnings in my  wrapper. And  I'm moving to opengl on windows because
avaloniaui now  integrates it .


 I think a top level exception handler and some workaround might be  a good
discovery because u can't ship with a leak like this.  I'm not using render
target from bk thread and if I even  did think i wanted to, I would do
context switch waits or  blocking in my own code, it shouldn't be in the
core because dispatchers are too nuanced.  Also it's unnecessary and
something to avoid, better to make a immutable copy , use producer/
consumer,  and avoid use of semaphores that depend on rough and widely
varying system timers, message queues.

**Comment by TechPizzaDev** [comment]
I'm not entirely sure, but it looks as if this null exception could have happened before my PR as well. Nonetheless, this should be fixable without sacrificing functionality.

And on the topic of deadlocks when exiting/finalizing, the GL device could dispose all queued actions on GraphicsDevice.Dispose, which would let all threads to continue (but in an unknown state since their operations will not have completed, although an exception could also be thrown that the device was disposed).

**Comment by MutsiMutsi** [comment]
I have the exact same issue but for me it happens rather consistently, I offload terrain generation off the main thread, most of that stuff has nothing to do with OpenGL and its perfectly fine to run off the main thread, stuff such as noise gen, however I do have a step within terrain gen that requires some GPU processing, so invoking that off of the main thread stuffs it in BlockOnUIThread awaiters, which cause deadlock when I exit the game, and subsequently await my terrain generation thread, this I can reproduce consistently. 

For me the super quick fix was defining a public static cancellationtoken within MG threading.cs and passing that into the awaiter.
```C#
public static class MonoGameUIBlockCancellationTokenHolder
{
    public static CancellationToken Token;
}

resetEvent.Wait(MonoGameUIBlockCancellationTokenHolder.Token);
```
I set that CT at startup of my game, and then I cancel that token when I shutdown my game, so that I can await my terrain gen thread to finish. Its quick, its dirty, it solved my problems. I'm just bumping this thread that this can be a consistent nuisance if you work with vanilla MG GL.

**Comment by ThomasFOG** [comment]
Closing this as DesktopGL is not designed as thread safe, nor is MonoGame actually. There is a tolerance to allow/trick resource loading from threads, but rendering or data fetch aren't supported nor planned to be threadable.
